rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/topic_cache.hpp

```diff
@@ -147,22 +186,26 @@ class TopicCache
     const std::string & topic_name,
     const std::string & type_name)
   {
-    if (topic_to_types_.find(topic_name) == topic_to_types_.end()) {
+    if (topic_name_to_topic_data_.find(topic_name) == topic_name_to_topic_data_.end()) {
       RCUTILS_LOG_DEBUG_NAMED(
         "rmw_fastrtps_shared_cpp",
         "unexpected removal on topic '%s' with type '%s'",
         topic_name.c_str(), type_name.c_str());
       return false;
     }
+    auto guid = iHandle2GUID(rtpsParticipantKey);
     {
-      auto & type_vec = topic_to_types_[topic_name];
-      type_vec.erase(std::find(type_vec.begin(), type_vec.end(), type_name));
+      auto & type_vec = topic_name_to_topic_data_[topic_name];
+      type_vec.erase(std::find_if(type_vec.begin(), type_vec.end(),
+        [type_name, guid](const auto & topic_info) {
+          return type_name.compare(std::get<1>(topic_info)) == 0 &&
+          guid == std::get<0>(topic_info);
```

Comment:
This is wrong as you are storing in guid the participant guid, which is the same for all the entities.